core/src/main/clojure/xtdb/expression/temporal.clj

```diff
@@ -374,33 +386,121 @@
   (defmethod expr/codegen-call [:+ t1 t2] [expr]
     (recall-with-flipped-args expr)))
 
-;;; subtract
-
-(defmethod expr/codegen-call [:- :timestamp-tz :timestamp-tz] [{[[_ x-unit _], [_ y-unit _]] :arg-types}]
-  (-> (with-arg-unit-conversion x-unit y-unit
-        #(do [:duration %]) #(do `(Math/subtractExact ~@%)))
+;; Subtraction
+
+;; DATE_DIFF implemented below - takes two datetimes (timestamp-local) with variable number of extra arguments:
+;; - 'start-field' - calculate difference in start-field and convert to truncated interval based on start-field 
+;;   ie, called for (DATEDIFF dt1 dt2 "HOUR")
+;; - 'start-field' & 'precision' - calculate difference in start-field, with an extra precision provided for seconds. 
+;;   Convert to truncated interval based on start-field with appropriate second precision.
+;;   ie, called for (DATEDIFF dt1 dt2 "SECOND (1, 3)")
+;; - 'start-field' & 'end-field' - calculate difference between the two in end-field and convert to truncated interval
+;;   with least significant field being end-field and most significant field being start-field.
+;;   ie, called for (DATEDIFF dt1 dt2 "DAY TO SECOND")
+;; - 'start-field', 'end-field' & precision - calculate difference between the two in end-field and convert to truncated interval, with
+;;   appropriate provided precision for seconds
+;;  ie, called for (DATEDIFF dt1 dt2 "DAY TO SECOND (6)")
+;; Some notes on how we behave based on values passed in:
+;; - We change return type of output based on fields passed in - if we start with YEAR or MONTH, return a `year-month`, otherwise a `month-day-nano`
+;; - When only time based fields entered (ie, HOUR, SECOND), we return everything in the Duration, calculated by our least significant field in the interval
+;;    - As we use `Duration` - we will always display in terms of HOURS, MINUTES and SECONDS
+(defn nanos-precision->duration [nanos precision] 
+  (let [second-nanos (bigdec (/ (long nanos) 1e9))
+        seconds (.setScale second-nanos 0 BigDecimal/ROUND_DOWN)
+        nanos (.longValueExact (.setScale (.multiply (.subtract second-nanos seconds) 1e9M) (- (long precision) 9) BigDecimal/ROUND_DOWN))]
+    (Duration/ofSeconds seconds nanos)))
+
+(defn date-diff-start-end-fields [x-unit y-unit start-field end-field precision] 
+  {:return-type (if (= ["YEAR" "MONTH"] [start-field end-field])
+                  [:interval :year-month]
+                  [:interval :month-day-nano])
+   :->call-code (fn [[x y]]
+                  (case [start-field end-field]
+                    ["YEAR" "MONTH"] `(PeriodDuration. (Period/of (.between ChronoUnit/YEARS ~(ts->ldt y y-unit) ~(ts->ldt x x-unit))
```

Comment:
Left out "MONTH" to "X" as it seemed a bit weird to represent (ie, how long is a MONTH in days?). Can add them in if we'd prefer - would probably need some particular math to decide on Months/Days based on the length of the months.